### PR TITLE
replace exit with a better shutdown mechanism

### PR DESCRIPTION
**Pull request by ldesgoui**
std::process::exit stops the process immediately, preventing custom
deconstructors from running

### CODE REVIEW

**Comment by RAnders00** [comment]
just wondering, what's the difference between a Notify and a oneshot::channel<()>?

**Comment by ldesgoui** [comment]
oneshot is one producer to one consumer, since any of the many websocket handlers can produce, it wouldn't work
mpsc exists (and even is in scope), we could use it instead

File: src/websocket.rs [hunk 1]
```diff
@@ -21,6 +21,7 @@ pub async fn listener(
     mut listener: TcpListener,
     tx_requested_follows: mpsc::Sender<(SocketAddr, Follows)>,
     tx_tweet: broadcast::Sender<Tweet>,
+    lifeline: &Arc<Notify>,
```
Comment (RAnders00) [inline src/websocket.rs:24]: wouldn't it be more consistent to accept `Arc<Notify>` in both methods, and not a reference in one and an owned type in the other?

File: src/websocket.rs [hunk 2]
```diff
@@ -60,6 +69,7 @@ async fn handler(
     addr: SocketAddr,
     tx_requested_follows: &mut mpsc::Sender<(SocketAddr, Follows)>,
     rx_tweet: broadcast::Receiver<Tweet>,
+    lifeline: Arc<Notify>,
```
Comment (RAnders00) [inline src/websocket.rs:72]: referring to this

File: src/websocket.rs (hunk 1 above)
Comment (ldesgoui) [reply src/websocket.rs:24]: it would just incur an additional unnecessary `Arc::clone`

Comment (RAnders00) [reply src/websocket.rs:24]: sorry, i had to read into it a bit more, I now see what's going on

**Pull request merged**